lib/flight_scheduler/message_processor.rb

```diff
@@ -140,20 +143,47 @@ def call(message)
       when 'JOB_CANCELLED'
         job_id = message[:job_id]
         Async.logger.info("Cancelling job:#{job_id}")
-        job_runner = FlightScheduler.app.job_registry.lookup_runner(job_id, 'BATCH')
-        if job_runner
-          job_runner.cancel
-          # The RUN_SCRIPT task will report back that the process has failed.
-          # We don't need to send any messages to the controller here.
+
+        # Deallocate the job to prevent any further job steps
+        FlightScheduler.app.job_registry.deallocate_job(job_id)
+
+        Async do |task|
+          # Allow other tasks to run a final time before cancelling
+          # This is a last attempt to collect any finished processes
+          task.yield
+
+          # Cancel all current runners
+          FlightScheduler.app.job_registry.lookup_runners(job_id).each do |_, runner|
+            runner.cancel
+          end
+
+          # Wait for the runners to finish and remove the job
+          task.yield until FlightScheduler.app.job_registry.lookup_runners(job_id).empty?
+          FlightScheduler.app.job_registry.remove_job(job_id)
+          MessageSender.send(command: 'NODE_DEALLOCATED', job_id: job_id)
+        end
+
+      when 'JOB_DEALLOCATED'
+        job_id = message[:job_id]
+        Async.logger.info("Deallocating job:#{job_id}")
+
+        # Deallocate the job to prevent any further job steps
+        FlightScheduler.app.job_registry.deallocate_job(job_id)
+
+        # Report back when all the runners have stop
+        Async do |task|
+          task.yield until FlightScheduler.app.job_registry.lookup_runners(job_id).empty?
+          FlightScheduler.app.job_registry.remove_job(job_id)
+          MessageSender.send(command: 'NODE_DEALLOCATED', job_id: job_id)
         end
-        FlightScheduler.app.job_registry.remove_job(job_id)
 
       else
         Async.logger.info("Unknown message #{message}")
       end
       Async.logger.debug("Processed message #{message.inspect}")
-    rescue
+    rescue => e
       Async.logger.warn("Error processing message #{$!.message}")
+      Async.logger.debug e.full_message
```

Comment:
And more pedantry regarding consistency with how we reference the exception.

Reply:
https://github.com/openflighthpc/flight-scheduler-daemon/pull/16/commits/a41a0979b1d080fef8a03bf02613ccbc01650e25